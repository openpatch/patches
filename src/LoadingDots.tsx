import { css } from "@emotion/react";
import { ReactNode } from "react";
import { Box, BoxProps } from "./Box";

export type LoadingDotsProps = {
  children?: ReactNode;
  size?: "small" | "medium" | "large";
  textColor?: BoxProps["textColor"];
};

export const LoadingDots = ({
  children,
  size = "medium",
  textColor,
}: LoadingDotsProps) => {
  let width = "0px";
  switch (size) {
    case "small": {
      width = "2px";
      break;
    }
    case "medium": {
      width = "4px";
      break;
    }
    case "large": {
      width = "6px";
      break;
    }
  }
  return (
    <Box
      css={css`
        @keyframes blink {
          /**
     * At the start of the animation the dot
     * has an opacity of .2
     */
          0% {
            opacity: 0.2;
          }
          /**
     * At 20% the dot is fully visible and
     * then fades out slowly
     */
          20% {
            opacity: 1;
          }
          /**
     * Until it reaches an opacity of .2 and
     * the animation can start again
     */
          100% {
            opacity: 0.2;
          }
        }

        & > span {
          /**
     * Use the blink animation, which is defined above
     */
          animation-name: blink;
          /**
     * The animation should take 1.4 seconds
     */
          animation-duration: 1.4s;
          /**
     * It will repeat itself forever
     */
          animation-iteration-count: infinite;
          /**
     * This makes sure that the starting style (opacity: .2)
     * of the animation is applied before the animation starts.
     * Otherwise we would see a short flash or would have
     * to set the default styling of the dots to the same
     * as the animation. Same applies for the ending styles.
     */
          animation-fill-mode: both;
        }

        & > span:nth-of-type(2) {
          /**
     * Starts the animation of the third dot
     * with a delay of .2s, otherwise all dots
     * would animate at the same time
     */
          animation-delay: 0.2s;
        }

        & > span:nth-of-type(3) {
          /**
     * Starts the animation of the third dot
     * with a delay of .4s, otherwise all dots
     * would animate at the same time
     */
          animation-delay: 0.4s;
        }
      `}
      display="inline-flex"
      alignItems="center"
    >
      {children && (
        <Box
          m="none"
          p="none"
          textColor={textColor}
          as="p"
          marginRight="xsmall"
        >
          {children}
        </Box>
      )}
      <Box
        backgroundColor={textColor || "currentColor"}
        borderRadius="full"
        width={width}
        marginRight="xxsmall"
        height={width}
        as="span"
      ></Box>
      <Box
        backgroundColor={textColor || "currentColor"}
        borderRadius="full"
        marginRight="xxsmall"
        width={width}
        height={width}
        as="span"
      ></Box>
      <Box
        backgroundColor={textColor || "currentColor"}
        borderRadius="full"
        width={width}
        height={width}
        as="span"
      ></Box>
    </Box>
  );
};
