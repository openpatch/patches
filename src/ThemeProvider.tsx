import React, { ReactElement } from "react";
import { ThemeProvider as EThemeProvider } from "emotion-theming";
import { Theme, base } from "./themes";
import { GlobalCssTheme } from "./themes/utils";

interface ThemeProviderProps {
  theme?: Theme;
  children: ReactElement;
}

const ThemeProvider = ({ theme = base, children }: ThemeProviderProps) => {
  return (
    <EThemeProvider theme={theme}>
      <GlobalCssTheme theme={theme} />
      {children}
    </EThemeProvider>
  );
};

export default ThemeProvider;
