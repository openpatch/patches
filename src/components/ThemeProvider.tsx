import React, { ReactElement } from "react";
import {
  ThemeProvider as SCThemeProvider,
  DefaultTheme,
} from "styled-components";
import defaultTheme from "../shared/theme";

interface ThemeProviderProps {
  theme?: DefaultTheme;
  children: ReactElement;
}

const ThemeProvider = ({
  theme = defaultTheme,
  children,
}: ThemeProviderProps) => {
  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};

export default ThemeProvider;
