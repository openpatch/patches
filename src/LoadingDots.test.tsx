import React from "react";
import { render } from "@testing-library/react";
import { LoadingDots, LoadingDotsProps } from "./LoadingDots";
import { ThemeProvider } from "./ThemeProvider";

const renderWithTheme = (props?: LoadingDotsProps) =>
  render(
    <ThemeProvider>
      <LoadingDots {...props} />
    </ThemeProvider>
  );

test("should render", () => {
  renderWithTheme();
});

test("should match snapshot", () => {
  const { container } = renderWithTheme();
  expect(container).toMatchSnapshot();
});
