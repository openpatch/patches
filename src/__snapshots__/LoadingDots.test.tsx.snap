// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
.emotion-0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0>span {
  -webkit-animation-name: blink;
  animation-name: blink;
  -webkit-animation-duration: 1.4s;
  animation-duration: 1.4s;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}

.emotion-0>span:nth-of-type(2) {
  -webkit-animation-delay: 0.2s;
  animation-delay: 0.2s;
}

.emotion-0>span:nth-of-type(3) {
  -webkit-animation-delay: 0.4s;
  animation-delay: 0.4s;
}

.emotion-1 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  background-color: #102A43;
  border-radius: 9999px;
  width: 4px;
  margin-right: 4px;
  height: 4px;
}

.emotion-2 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  background-color: #102A43;
  border-radius: 9999px;
  margin-right: 4px;
  width: 4px;
  height: 4px;
}

.emotion-3 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  background-color: #102A43;
  border-radius: 9999px;
  width: 4px;
  height: 4px;
}

<div>
  <div
    class="emotion-0"
    display="inline-flex"
  >
    <span
      class="emotion-1"
      height="4px"
      width="4px"
    />
    <span
      class="emotion-2"
      height="4px"
      width="4px"
    />
    <span
      class="emotion-3"
      height="4px"
      width="4px"
    />
  </div>
</div>
`;
